### PR TITLE
Pull request for due-adc to release

### PR DESCRIPTION
**Pull request by 56KbModem**
The signal collecting logic is implemented.
I hereby make this pull-request to the release branch.

Please review.

### CODE REVIEW

File: code/src/microphone_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    for (int i=0; i < 63; ++i){
+        sample_buffer[i] = microphone_controller_c::read_mic(); // fill sample buffer
+    }
+}
+
+uint16_t* r2d2::microphone::microphone_controller_c::read_buffer(){
+    return sample_buffer; // return pointer to buffer
+}
+
+void r2d2::microphone::microphone_controller_c::mute() {
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: Why do you return here?

File: code/src/microphone_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    if (!muted){ // if the mic is not muted
+        muted = true; // mute the microphone
+    }
+    return;
+    
+}
+
+void r2d2::microphone::microphone_controller_c::unmute(){
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: Why do you return here?

File: code/src/microphone_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    if (!muted){
+        return r2d2::microphone::microphone_controller_c::mic.read(); // return adc pin (microphone) value
+    }
+    else{
+        return 0; // don't return an audible value
+    }
+}
+
+void r2d2::microphone::microphone_controller_c::sample(){
+    for (int i=0; i < 63; ++i){
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: This 63 is hardcoded. Try and use the buffer_length you use in the module_c here

File: code/src/microphone_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        sample_buffer[i] = microphone_controller_c::read_mic(); // fill sample buffer
+    }
+}
+
+uint16_t* r2d2::microphone::microphone_controller_c::read_buffer(){
+    return sample_buffer; // return pointer to buffer
+}
+
+void r2d2::microphone::microphone_controller_c::mute() {
+    if (!muted){ // if the mic is not muted
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: I don't think the if-statement here is needed. Just set the pin, doesn't matter if it is already muted or not

File: code/src/microphone_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        muted = true; // mute the microphone
+    }
+    return;
+    
+}
+
+void r2d2::microphone::microphone_controller_c::unmute(){
+    if (muted){ // if the mic is muted
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: I don't think the if-statement here is needed. Just set the pin, doesn't matter if it is already muted or not

File: code/main.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#include <hwlib.hpp>
+
+#include <comm.hpp>
+#include "headers/module.hpp"
```
Comment (Official12345) [inline code/main.cpp]: you can just include "module.hpp" if im correct

File: code/main.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#include "headers/microphone_controller.hpp"
```
Comment (Official12345) [inline code/main.cpp]: you can just include "microphone_controller.hpp" if im correct

File: code/src/microphone_controller.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
+#include "../headers/microphone_controller.hpp"
```
Comment (Official12345) [inline code/src/microphone_controller.cpp]: you can just include "microphone_controller.hpp" if im correct

**Review by Official12345 — CHANGES_REQUESTED** [review]
Left comments at the line numbers

File: code/src/microphone_controller.cpp (hunk 1 above)
Comment (56KbModem) [reply code/src/microphone_controller.cpp]: I do not, it is a void function which changes the private "muted" variable

File: code/src/microphone_controller.cpp (hunk 2 above)
Comment (56KbModem) [reply code/src/microphone_controller.cpp]: I do not, it is a void function which changes the private "muted" variable

File: code/src/microphone_controller.cpp (hunk 3 above)
Comment (56KbModem) [reply code/src/microphone_controller.cpp]: The 63 is the internal buffer size for internal communication (in 16 bit chunks) as agreed upon with @itzandroidtab . However I did have this magic number replaced with a get_buffer_size() function which now makes it so the magic number is only in one place with commentary for why it is statically defined.

File: code/src/microphone_controller.cpp (hunk 5 above)
Comment (56KbModem) [reply code/src/microphone_controller.cpp]: agreed, changed that now

File: code/src/microphone_controller.cpp (hunk 4 above)
Comment (56KbModem) [reply code/src/microphone_controller.cpp]: agreed

**Review by Official12345 — APPROVED** [review]

**Pull request merged**